sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+
+[source,c++]
+----
+id_type get_group_linear_id() const;
+----
+_Returns_: A linearized version of the `id` returned by `get_group_id()`.
+
+[source,c++]
+----
+id_type get_local_linear_id() const;
+----
+_Returns_: A linearized version of the `id` returned by `get_local_linear_id()`.
+
+[source,c++]
+----
+range_type get_group_linear_range() const;
+----
+_Returns_: A linearized version of the `id` returned by `get_group_range()`.
+
+[source,c++]
+----
+range_type get_local_linear_range() const;
+----
+_Returns_: A linearized version of the `id` returned by `get_local_range()`.
+
+[source,c++]
+----
+bool leader() const;
+----
+_Returns_: `true` for exactly one work-item in the cluster-group, if the calling
+work-item is the leader of the cluster-group, and `false` for all other
+work-items in the cluster-group. The leader of the cluster-group is guaranteed
+to be the work-item for which `get_local_id()` returns 0.
+
+
+==== Usage Examples
+
+A `cluster_group` can be used to apply group algorithms to subsets of data:
+
+[source, c++]
+----
+auto sg = it.get_sub_group();
+
+// reduce over contiguous groups of 8 elements
+auto cluster = sycl::ext::oneapi::experimental::get_cluster_group<8>(sg);
+auto result = sycl::reduce_over_group(cluster, buf[it.get_local_linear_id()], sycl::plus<>());
+
+// write result out once per group
+if (cluster.leader()){
+  buf[cluster.get_group_id()] = result;
+}
+----
+
+A `cluster_group` can be used to provide an interface accepting a specific
+number of work-items:
+
+[source, c++]
+----
+void func_that_needs_4_work_items(sycl::ext::oneapi::experimental::cluster_group<4> group);
+----
+
+
+=== Tangle-Groups
+
+A tangle-group is a non-contiguous subset of a group representing work-items
+executing in a tangle. A tangle-group can therefore be used to capture all
+work-items currently executing the same control flow.
+
+
+==== Creation
+
+New tangle-groups are created by partitioning an existing group, using the
+`get_tangle_group` free-function.
+
+NOTE: Creating a tangle-group may require synchronization of the parent group
+on some devices, since it may be necessary to wait for work-items to
+reconverge. For consistency, synchronization is required by all
+implementations.
+
+[source, c++]
+----
+namespace ext::oneapi::experimental {
+
+template <typename Group>
+tangle_group get_tangle_group(Group group);
+
+} // namespace ext::oneapi::experimental
+----
+
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>> &&
+std::is_same_v<Group, sycl::sub_group>` is true.
+
+_Preconditions_: All work-items in `group` must encounter this function in
+converged control flow.
+
+_Effects_: Synchronizes all work-items in `group`.
+
+_Returns_: A `tangle_group` consisting of the work-items in `group` which are
+part of the same tangle.
+
+
+==== `tangle_group` Class
+
+[source, c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+class tangle_group {
```

Comment:
Why isn't this class templated on the parent group type?

Reply:
Fixed as part of e34728ab6.